af/fastpath/fastpath/core.py

```diff
@@ -0,0 +1,574 @@
+                    log.debug("matched body fp %s %r", msm_cc, bm)
+
+                per_s("fingerprints_bytes", len(body), tb)
+
+        del (body)
+
+        # Match HTTP headers if found
+        headers = r.get("headers", {})
+        if not headers:
+            continue
+        headers = {h.lower(): v for h, v in headers.items()}
+        for fp in fingerprints[msm_cc].get("header_full", []):
+            name = fp["header_name"]
+            if name in headers and headers[name] == fp["header_full"]:
+                matches.append(fp)
+                log.debug("matched header full fp %s %r", msm_cc, fp["header_full"])
+
+        for fp in fingerprints[msm_cc].get("header_prefix", []):
+            name = fp["header_name"]
+            prefix = fp["header_prefix"]
+            if name in headers and headers[name].startswith(prefix):
+                matches.append(fp)
+                log.debug("matched header prefix %s %r", msm_cc, prefix)
+
+    return matches
+
+
+@metrics.timer("score_measurement")
+def score_measurement(msm, matches):
+    """Calculate measurement scoring. Returns a summary dict
+    """
+    # Extract interesting fields
+    fields = (
+        "input",
+        "measurement_start_time",
+        "probe_asn",
+        "probe_cc",
+        "report_id",
+        "test_name",
+        "test_start_time",
+    )
+    summary = {k: msm[k] for k in fields}
+
+    # Blocking locality: global > country > ISP > local
+    # unclassified locality is stored in "blocking_general"
+
+    scores = {f"blocking_{l}": 0.0 for l in LOCALITY_VALS}
+
+    for m in matches:
+        l = "blocking_" + m["locality"]
+        scores[l] += 1.0
+
+    # Feature extraction
+
+    if msm.get("title_match", True) is False:
+        scores["blocking_general"] += 0.5
+
+    if "body_proportion" in msm["test_keys"]:
+        # body_proportion can be missing or can be None
+        delta = abs((msm["test_keys"]["body_proportion"] or 1.0) - 1.0)
+        scores["blocking_general"] += delta
+
+    # TODO: add IM tests scoring here
+
+    # TODO: add heuristic to split blocking_general into local/ISP/country/global
+    scores["blocking_general"] += (
+        scores["blocking_country"]
+        + scores["blocking_global"]
+        + scores["blocking_isp"]
+        + scores["blocking_local"]
+    )
+    summary["scores"] = scores
+    return summary
+
+
+@metrics.timer("trivial_id")
+def trivial_id(msm):
+    """Generate a trivial id of the measurement to allow upsert if needed
+    - 32-bytes hexdigest
+    - Deterministic / stateless with no DB interaction
+    - Malicious/bugged msmts with collisions on report_id/input/test_name lead
+    to different hash values avoiding the collision
+    - Malicious/duplicated msmts that are semantically identical to the "real"
+    one lead to harmless collisions
+    """
+    # NOTE: we want the id to stay the same when a msmt is fetched from SSH
+    # and from a can on ooni-data-private/canned
```

Comment:
I believe this will be the case, though beware that if instead of using canned data you use autoclaved data the hash of the JSON will not match, because some extra keys are added in autoclaving (for example the `bucket_date`).

Reply:
Looking at the code more I see that for `.yaml` measurements you are calling `normalize_entry` and passing it a fixed `bucket_date` or `2018-05-07` (maybe this value should be more clearly something hardcoded like a non-existent bucket_date) and the filename is set to `??`.

Reply:
Updated in the last commit. Added testing and temporarily disabled handling YAML.